File: bitbots_transform/scripts/base_footprint.py
#!/usr/bin/env python2.7
import tf2_ros
from tf import transformations
import rospy
from sensor_msgs.msg import Imu
from geometry_msgs.msg import TransformStamped


class BaseFootprintBroadcaster:
    def __init__(self):
        rospy.init_node("base_footprint_broadcaster")
        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
        self.tf_buffer = tf2_ros.Buffer(cache_time=rospy.Duration(2.0))
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)

        rospy.Subscriber("imu/data", Imu, self.callback_imu)
        self.tf = TransformStamped()
        self.tf.header.frame_id = "base_link"
        self.tf.child_frame_id = "base_footprint"
        self.tf.transform.translation.x = 0.0
        self.tf.transform.translation.y = 0.0

        rospy.spin()

    def callback_imu(self, msg):
        self.tf.header.stamp = msg.header.stamp
        try:
            tf_right = self.tf_buffer.lookup_transform("base_link", "r_sole", msg.header.stamp, timeout=rospy.Duration(0.1))
            tf_left = self.tf_buffer.lookup_transform("base_link", "l_sole", msg.header.stamp, timeout=rospy.Duration(0.1))
        except tf2_ros.LookupException:
            rospy.logwarn("still waiting for transforms")
            return

        if tf_right.transform.translation.z < tf_left.transform.translation.z:
            self.tf.transform.translation.z = tf_right.transform.translation.z

        else:
            self.tf.transform.translation.z = tf_right.transform.translation.z

        self.tf.transform.rotation = msg.orientation

        """
        self.tf.transform.rotation.x = msg.orientation.x
        self.tf.transform.rotation.y = msg.orientation.y
        self.tf.transform.rotation.z = msg.orientation.z
        self.tf.transform.rotation.w = msg.orientation.w"""


        self.tf_broadcaster.sendTransform(self.tf)


if __name__ == "__main__":
    BaseFootprintBroadcaster()

File: bitbots_transform/src/base_footprint.cpp


#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <std_msgs/Time.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/transform_listener.h>
#include <geometry_msgs/TransformStamped.h>


class BaseFootprintBroadcaster
{
public:
    BaseFootprintBroadcaster();
private:
    //void imu_callback(const sensor_msgs::Imu::ConstPtr& imu_msg);
    //ros::Subscriber imu_subscriber;
    geometry_msgs::TransformStamped tf;
    tf2_ros::Buffer tfBuffer;
    tf2_ros::TransformListener tfListener;
    //bool imu;
};

BaseFootprintBroadcaster::BaseFootprintBroadcaster() : tfBuffer(ros::Duration(10.0)),  tfListener(tfBuffer)
{
    //setup tf listener and broadcaster
    //register callback

    ros::NodeHandle n("~");
    //std::string imu_topic;
    //n.getParam("imu_topic", imu_topic);
    //imu_subscriber = n.subscribe(imu_topic, 1, &BaseFootprintBroadcaster::imu_callback, this);

    

    tf = geometry_msgs::TransformStamped();
    tf.header.frame_id = "base_link";
    tf.child_frame_id = "base_footprint";
    tf.transform.translation.x = 0.0;

    static tf2_ros::TransformBroadcaster br;
    ros::Rate r(30.0);
    while(ros::ok())
    {
        /*ros::spinOnce();
        if(!imu)
        {
            tf.header.stamp = ros::Time::now();
        }*/

        geometry_msgs::TransformStamped tf_right, tf_left;
        try{
            tf_right = tfBuffer.lookupTransform("base_link", "r_sole", ros::Time::now(), ros::Duration(0.1));
            tf_left = tfBuffer.lookupTransform("base_link", "l_sole",  ros::Time::now(), ros::Duration(0.1));
        }catch(...){
            continue;
        }
        // check which foot is support foot (which foot is on the ground)
        if(tf_right.transform.translation.z < tf_left.transform.translation.z) {
            tf.transform.translation.z = tf_right.transform.translation.z;
            tf.transform.rotation = tf_right.transform.rotation;
            tf.header.stamp = tf_right.header.stamp;
        } else {
            tf.transform.translation.z = tf_left.transform.translation.z;
            tf.transform.rotation = tf_left.transform.rotation;
            tf.header.stamp = tf_left.header.stamp;
        }

        tf.transform.translation.y = (tf_left.transform.translation.y + tf_right.transform.translation.y)/2;

        br.sendTransform(tf);
        r.sleep();

    }
    ros::spin();
}

/*void BaseFootprintBroadcaster::imu_callback(const sensor_msgs::Imu::ConstPtr& imu_msg)
{

    imu = true;

    if((imu_msg->orientation.x + imu_msg->orientation.y + imu_msg->orientation.z + imu_msg->orientation.w) == 0.0)
    {
        ROS_WARN("imu message has no orientation set, use complementary filter to get one ");
        return;
    }
    tf.transform.rotation = imu_msg->orientation;


}*/

int main(int argc, char **argv)
{
    ros::init(argc, argv, "base_footprint");

    BaseFootprintBroadcaster b;
}

